### PR TITLE
Email verification

### PR DESCRIPTION
**Pull request by Mari-Hayashi**
Verify user's email when they sign up.

### CODE REVIEW

**Comment by Mari-Hayashi** [comment]
@johan1505  Hi Johan, I haven't implemented the front end yet, but could you quickly review if the back end design looks good to you?

I added a boolean field "emailVerified" in the user entity.

Initially, when a user creates an account, emailVerified is set to false. The website sends an email to the user with the verification link. (Look at createUser in src/server/src/modules/user.ts)

The user cannot log in when emailVerified is false. (Look at user/sign-in API)

When the user follows the link in the email, the website calls a backend API api/user/verify and verifies their email (and set emailVerified to true).

This is how the verification email looks like:
image
(Maybe we want to add new lines between sentences.)

Please let me know what you think!

File: src/server/src/modules/user.ts [hunk 1]
```diff
@@ -48,8 +51,32 @@ export const createUser = (
...
+        verificationKey: verificationKeyString
+    });
+
+    await verificationKey.save();
+    
+    // Send email with the verification link.
+    sendEmail(
+        email,
+        `Verify your email address`,
+        `Welcome to R'match! Please verify your email address so we know that it's you. http://localhost:3000/verify/${verificationKeyString}`,
```
Comment (johan1505) [inline src/server/src/modules/user.ts]: This is fine for development; however, in production, the link should start with https://obscure-ocean-12960.herokuapp.com/. So, the content of the link will depend on this value `process.env.NODE_ENV` if that value equals to "development" or "test" use localhost, if it equals to "production" use the heroku link

File: src/server/src/entities/verificationKey.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    @Column()
+    userId: number;
+
+    @OneToOne(() => User)
+    @JoinColumn()
+    user: User;
+
+    @Column()
+    verificationKey: string;
```
Comment (johan1505) [inline src/server/src/entities/verificationKey.ts]: Kind of redundant for this column to have the same name as the table. Maybe just use "key"?

File: src/server/src/entities/user.ts [hunk 3]
```diff
@@ -1,5 +1,14 @@
...
+    Entity,
+    Column,
+    PrimaryGeneratedColumn,
+    OneToMany,
+    JoinColumn,
+    OneToOne,
+    BaseEntity
+} from 'typeorm';
 import { Message } from './message';
+import { VerificationKey } from './verificationKey';
```
Comment (johan1505) [inline src/server/src/entities/user.ts]: I think you forgot to use this entity

File: src/server/src/modules/user.ts (hunk 1 above)
Comment (johan1505) [inline src/server/src/modules/user.ts]: The body of the email could be something like this:

Welcome to R'Match!

Your account has been created. Please follow this link: *your link* to verify your email and complete your registration.

File: src/server/src/routes/user/index.ts [hunk 4]
```diff
@@ -5,20 +5,27 @@ import passport from 'passport';
...
 import { errors } from '@shared/errors';
 import logger from '@shared/Logger';
-import { findUserByEmail, registerUser } from '@modules/user';
+import { findUserByEmail, registerUser, verifyEmail } from '@modules/user';
 import { User } from '@entities/user';
 import { signToken } from '@lib/jwt';
 import { validationMiddleware } from '@middlewares/validation';
-import { signUpSchema, signInSchema } from './schemas';
+import { signUpSchema, signInSchema, emailVerificationSchema } from './schemas';
+import { verify } from 'jsonwebtoken';
```
Comment (johan1505) [inline src/server/src/routes/user/index.ts]: I don't think this import is needed

**Review by johan1505 — COMMENTED** [review]
I left some comments. The overall design is great, I like it :)

File: src/server/src/modules/user.ts (hunk 1 above)
Comment (Mari-Hayashi) [reply src/server/src/modules/user.ts]: Done.

File: src/server/src/entities/verificationKey.ts (hunk 2 above)
Comment (Mari-Hayashi) [reply src/server/src/entities/verificationKey.ts]: Done.

File: src/server/src/entities/user.ts (hunk 3 above)
Comment (Mari-Hayashi) [reply src/server/src/entities/user.ts]: Done.

File: src/server/src/modules/user.ts (hunk 1 above)
Comment (Mari-Hayashi) [reply src/server/src/modules/user.ts]: Done.

File: src/server/src/routes/user/index.ts (hunk 4 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/user/index.ts]: Done.

**Comment by Mari-Hayashi** [comment]
@johan1505 Hi! This PR is ready for review. (The gif starts from where I'm creating an account.)
EmailVerification

**Pull request merged**